### PR TITLE
#144 add option to use basic token auth method

### PR DESCRIPTION
**Pull request by richardjq**
PR to fix [#144](https://github.com/jenkinsci/oic-auth-plugin/issues/114) 

For manual configuration:
Add an additional configuration entry for Token Auth Method that allows the user to select whether to use basic or post for the token client auth.

For auto configuration:
Inspect `token_endpoint_auth_methods_supported` from the OpenId configuration endpoint and use one of the available options.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by jeremymcgee73** [comment]
This would be awesome to get merged in. This solves the bootstrapping problem.

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 1]
```diff
@@ -148,13 +150,15 @@ public OicSecurityRealm(String clientId, String clientSecret, String wellKnownOp
...
             this.tokenServerUrl = config.getTokenEndpoint();
+            this.tokenAuthMethod = config.getPreferredTokenAuthMethod();
             this.userInfoServerUrl = config.getUserinfoEndpoint();
             this.scopes = config.getScopesSupported() != null && !config.getScopesSupported().isEmpty() ? StringUtils.join(config.getScopesSupported(), " ") : "openid email";
             this.logoutFromOpenidProvider = logoutFromOpenidProvider != null;
            	this.endSessionEndpoint = config.getEndSessionEndpoint();
         } else {
             this.authorizationServerUrl = authorizationServerUrl;
             this.tokenServerUrl = tokenServerUrl;
+            this.tokenAuthMethod = tokenAuthMethod;
```
Comment (michael-doubez) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:161]: Wouldn't that break previous configuration ?
Default to "client_secret_post" if empty or null.

File: src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/help-tokenAuthMethod.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<div>
+  Required. The supported token endpoint authentication method to use.
```
Comment (michael-doubez) [inline src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/help-tokenAuthMethod.html:2]: List supported methods.

File: src/test/resources/org/jenkinsci/plugins/oic/ConfigurationAsCode.yml [hunk 3]
```diff
@@ -14,5 +14,6 @@ jenkins:
       groupsFieldName: groupsFieldName
       logoutFromOpenidProvider: true
       scopes: scopes
+      tokenAuthMethod: client_secret_post
```
Comment (michael-doubez) [inline src/test/resources/org/jenkinsci/plugins/oic/ConfigurationAsCode.yml:17]: Add test case for wrong "tokenAuthMethod" value.

**Review by michael-doubez — CHANGES_REQUESTED** [review]
In the case of config-as-code (or in case of code change), using string the the logic implmented  is error prone; any typing error won't be detected and no error issued to help the user.

I'd rather you define an enum with string mapping.
It is a bother but users will bless you in the long run.

File: src/test/resources/org/jenkinsci/plugins/oic/ConfigurationAsCode.yml (hunk 3 above)
Comment (richardjq) [reply src/test/resources/org/jenkinsci/plugins/oic/ConfigurationAsCode.yml:17]: I can find no examples that do this for CasC. If you show me how, I would be glad to add it.

**Review by michael-doubez — APPROVED** [review]
LGTM

**Pull request merged**